### DISCUSSION TITLE
(Approximated) Multiway Cut and k-Cut

### DISCUSSION THREAD

**Discussion by atomassi** [networkx/networkx#4530] [Algorithms]
Hi all,
there are two problems involving natural generalizations of the minimum cut problem which I believe might be of interest to be added.  

Let G = (V, E) be an undirected graph.

#### Minimum Multiway Cut

Given a set of terminals S = {s1, s2, . . . , sk } ⊆ V , a multiway cut is a set of edges whose removal disconnects the terminals from each other. The min-multiway cut problem asks for the minimum weight multiway cut.   
Note that the case in which k=2 corresponds to the _minimum s-t cut_ and can be solved optimally in polynomial time. For k≥ 3, it’s NP-hard [[2]](#2).

#### Minimum k-Cut
A k-cut is a set of edges whose removal leaves k connected components. The  min-k-cut problem asks for a minimum weight k-cut.   
Note that the case in which k=2 corresponds to the _global minimum cut_ and can be solved optimally in polynomial time. For k≥ 3, it’s NP-hard [[1]](#1). 

## Proposed Algorithms

| Problem    	        | Approximation Ratio	|  Reference	| 
|---------------	|:-----:	                         |:-------:	|
| Multiway Cut 	| 2-2/k 	                         |   [[1]](#1) - Chapter 4	|  
| k-Cut   	                 | 2-2/k  	                         |   [[1]](#1) - Chapter 4	|    

Do you think NetworkX would benefit from them being added to the _approximation_ module? 

## References
<a id="1">[1]</a> 
Vazirani, Vijay V. *Approximation algorithms*. Springer Science & Business Media, 2013.

<a id=“2”>[2]</a> 
Dahlhaus, Elias, et al. *The complexity of multiway cuts.*  Proceedings of the twenty-fourth annual ACM symposium on Theory of computing. 1992

**Comment by dschult** [comment]
Yes, ythat would be of interest.  

We have one kcutsets function in `algorithms/connectivity/kcutsets.py`
I think the docs for that could be improved, but more to the point, these approximation algorithms should go in either `algorithms/approximation/maxcut.py` or as a new module called `algorithms/approximation/kcutsets.py`.  If there is a better name that's fine too, but these match other parts of the package.

If you have any questions about how to organize or set things up let us know. :)

**Reply by atomassi** [reply]
@dschult I've just noticed that I forgot to update this thread with the PR #4534 (link given by [reference](https://github.com/networkx/networkx/pull/4534). )
